javascript/ql/lib/semmle/javascript/frameworks/Handlebars.qll

```diff
@@ -27,3 +27,126 @@ module Handlebars {
     SafeString() { this = any(Handlebars h).getAConstructorInvocation("SafeString") }
   }
 }
+
+/** Provides logic for taint steps for the handlebars library. */
+private module HandlebarsTaintSteps {
+  /**
+   * Gets a reference to a compiled Handlebars template.
+   */
+  private DataFlow::SourceNode compiledTemplate(DataFlow::CallNode compileCall) {
+    result = compiledTemplate(DataFlow::TypeTracker::end(), compileCall)
+  }
+
+  private DataFlow::SourceNode compiledTemplate(
+    DataFlow::TypeTracker t, DataFlow::CallNode compileCall
+  ) {
+    t.start() and
+    result = any(Handlebars::Handlebars hb).getAMethodCall(["compile", "template"]) and
+    result = compileCall
+    or
+    exists(DataFlow::TypeTracker t2 | result = compiledTemplate(t2, compileCall).track(t2, t))
+  }
+
+  private predicate isHelperParam(
+    string helperName, DataFlow::FunctionNode helperFunction, DataFlow::ParameterNode param,
+    int paramIndex
+  ) {
+    exists(DataFlow::CallNode registerHelperCall |
+      registerHelperCall = any(Handlebars::Handlebars hb).getAMethodCall("registerHelper") and
+      registerHelperCall.getArgument(0).mayHaveStringValue(helperName) and
+      helperFunction = registerHelperCall.getArgument(1).getAFunctionValue() and
+      param = helperFunction.getParameter(paramIndex)
+    )
+  }
+
+  /** Holds if `call` is a block wrapped inside curly braces inside the template `templateText`. */
+  bindingset[templateText]
+  predicate templateHelperParamBlock(string templateText, string call) {
+    call = templateText.regexpFind("\\{\\{[^}]+\\}\\}", _, _).regexpReplaceAll("[{}]", "").trim() and
+    call.regexpMatch(".*\\s.*")
+  }
+
+  /**
+   * Holds for calls to helpers from handlebars templates.
+   *
+   * ```javascript
+   * hb.compile("contents of file {{path}} are: {{catFile path}} {{echo p1 p2}}");
+   * ```
+   *
+   * In the example, the predicate will hold for:
+   *
+   * * helperName="catFile", argIdx=1, arg="path"
+   * * helperName="echo", argIdx=1, arg="p1"
+   * * helperName="echo", argIdx=2, arg="p2"
+   *
+   * The initial `{{path}}` will not be considered, as it has no arguments.
+   */
+  bindingset[templateText]
+  private predicate isTemplateHelperCallArg(
```

Comment:
```suggestion
  private predicate isTemplateWithHelperCall(
``` 

A fully descriptive name might be something like: `isTemplateWithHelperCallThatHasArgument`, but that would be unwieldy.

Reply:
true, a bit too unwieldy imo. I'll leave it. 👍